sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -266,47 +267,75 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
       relation,
       projects,
       filterPredicates,
-      (requestedColumns, pushedFilters) => {
-        scanBuilder(requestedColumns, selectFilters(pushedFilters).toArray)
+      (requestedColumns, _, pushedFilters) => {
+        scanBuilder(requestedColumns, pushedFilters.toArray)
       })
   }
 
-  // Based on Catalyst expressions.
+  // Based on Catalyst expressions. The `scanBuilder` function accepts three arguments:
+  //
+  //  1. A `Seq[Attribute]`, containing all required column attributes, used to handle traits like
+  //     `PrunedFilteredScan`.
+  //  2. A `Seq[Expression]`, containing all gathered Catalyst filter expressions, used by
+  //     `CatalystScan`.
+  //  3. A `Seq[Filter]`, containing all data source `Filter`s that are converted from (possibly a
+  //     subset of) Catalyst filter expressions and can be handled by `relation`.
```

Comment:
The first `Seq[Expression]` argument is only used to handle `CatalystScan`, which is only left for experimenting purposes, no built-in concrete data sources implement `CatalystScan` now. The second `Seq[Filter]` argument is used to handle all other relation traits that support filter push-down, e.g. `PrunedFilteredScan` and `HadoopFsRelation`. Added comments to explain this.